### PR TITLE
hc filter: treat degraded hosts as healthy in non-passthrough mode

### PR DESCRIPTION
**Pull request by snowp**
Updates the health check filter to treat degraded hosts as healthy for
the purpose of computing the "health" of each cluster. This captures the
idea that these hosts are still available for routing: if the upstream
cluster is fully degraded it's still available, so there shouldn't be a
need for Envoy to declare itself unhealthy.

Signed-off-by: Snow Pettersen <snowp@squareup.com>

*Risk Level*: Low, small change to behavior
*Testing*: Added UT
*Docs Changes*: Updated filter docs + proto
*Release Notes*: n/a

### CODE REVIEW

**Comment by snowp** [comment]
My thinking is that this seems like a natural change and doesn't need a config option, but if someone thinks otherwise I'd be happy to put it behind a flag

**Comment by junr03** [comment]
lgtm codewise. 

I am not aware of how fully people are using degraded host semantics other than you all at square @snowp, so I am not sure if this warrants a config option. 

However, If we are being completely hard about the rules, this is a change in previous behavior, and thus should undergo either a config option with the previous behavior as default or a deprecation cycle.

Does any other @envoyproxy/maintainers have an opinion?

**Comment by mattklein123** [comment]
> Does any other @envoyproxy/maintainers have an opinion?

I think it's OK to not have a config option here as long as it's documented.

File: docs/root/intro/arch_overview/health_checking.rst [hunk 1]
```diff
@@ -92,7 +92,7 @@ operation:
 * **No pass through, computed from upstream cluster health**: In this mode, the health checking
   filter will return a 200 or a 503 depending on whether at least a :ref:`specified percentage
   <envoy_api_field_config.filter.http.health_check.v2.HealthCheck.cluster_min_healthy_percentages>` of the
-  servers are healthy in one or more upstream clusters. (If the Envoy server is in a draining
+  servers are available in one or more upstream clusters. (If the Envoy server is in a draining
```
Comment (junr03) [inline docs/root/intro/arch_overview/health_checking.rst]: I suggest: available (healthy + degraded)

File: source/extensions/filters/http/health_check/health_check.cc [hunk 2]
```diff
@@ -132,10 +132,11 @@ void HealthCheckFilter::onComplete() {
           }
         }
         // In the general case, consider the service unhealthy if fewer than the
-        // specified percentage of the servers in the cluster are healthy.
+        // specified percentage of the servers in the cluster are available.
```
Comment (junr03) [inline source/extensions/filters/http/health_check/health_check.cc]: same here. although perhaps less necessary as the sum is right below

**Pull request merged**